crates/cow-amm/src/maintainers.rs

```diff
@@ -0,0 +1,80 @@
+use {
+    crate::{cache::Storage, Amm},
+    contracts::ERC20,
+    ethcontract::futures::future::{join_all, select_ok},
+    ethrpc::Web3,
+    shared::maintenance::Maintaining,
+    std::sync::Arc,
+    tokio::sync::RwLock,
+};
+
+pub struct EmptyPoolRemoval {
+    storage: Arc<RwLock<Vec<Storage>>>,
+    web3: Web3,
+}
+
+impl EmptyPoolRemoval {
+    pub fn new(storage: Arc<RwLock<Vec<Storage>>>, web3: Web3) -> Self {
+        Self { storage, web3 }
+    }
+
+    /// Checks if the given AMM has a zero token balance.
+    async fn check_single(&self, amm: Arc<Amm>) -> bool {
+        let amm_address = amm.address();
+        let futures = amm
+            .traded_tokens()
+            .iter()
+            .map(move |token| async move {
+                ERC20::at(&self.web3, *token)
+                    .balance_of(*amm_address)
+                    .call()
+                    .await
+                    .map_err(|err| {
+                        tracing::warn!(
+                            amm = ?amm_address,
+                            ?token,
+                            ?err,
+                            "failed to check AMM token balance"
+                        );
+                    })
+                    .and_then(|balance| if balance.is_zero() { Ok(()) } else { Err(()) })
+            })
+            .map(Box::pin);
+        select_ok(futures).await.is_ok()
```

Comment:
Agree that the reverse logic is not straight forward. Let's add a comment:

```suggestion
        // If any future resolved to Ok(()), then there exists a non-zero balance.
        select_ok(futures).await.is_ok()
```